pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotEvaluateLiteralRule.java

```diff
@@ -0,0 +1,272 @@
+import org.apache.pinot.common.function.FunctionInvoker;
+import org.apache.pinot.common.function.FunctionRegistry;
+import org.apache.pinot.spi.utils.TimestampUtils;
+import org.apache.pinot.sql.parsers.SqlCompilationException;
+
+
+/**
+ * SingleValueAggregateRemoveRule that matches an Aggregate function SINGLE_VALUE and remove it
+ *
+ */
+public class PinotEvaluateLiteralRule extends RelOptRule {
+  public static final PinotEvaluateLiteralRule INSTANCE =
+      new PinotEvaluateLiteralRule(PinotRuleUtils.PINOT_REL_FACTORY);
+
+  public PinotEvaluateLiteralRule(RelBuilderFactory factory) {
+    super(operand(RelNode.class, any()), factory, null);
+  }
+
+  @Override
+  public boolean matches(RelOptRuleCall call) {
+    // Traverse the relational expression using a RexShuttle visitor
+    AtomicBoolean hasEvaluateCall = new AtomicBoolean(false);
+    call.rel(0).accept(new RexShuttle() {
+      @Override
+      public RexNode visitCall(RexCall call) {
+        // Check if all operands are RexLiteral
+        if (call.operands.stream().allMatch(operand -> (operand instanceof RexLiteral))) {
+          // If all operands are literals, this call can be evaluated
+          hasEvaluateCall.set(true);
+        }
+        call.operands.forEach(operand -> {
+          if (hasEvaluateCall.get()) {
+            // Early terminate if we already found one evaluate call
+            return;
+          }
+          // Recursively call visitCall on all RexCall operands
+          if (operand instanceof RexCall) {
+            visitCall((RexCall) operand);
+          }
+        });
+        // Return the call to continue traversal
+        return call;
+      }
+    });
+    return hasEvaluateCall.get();
+  }
+
+  @Override
+  public void onMatch(RelOptRuleCall call) {
+    RelNode rootRelNode = call.rel(0);
+    RexBuilder rexBuilder = rootRelNode.getCluster().getRexBuilder();
+    Map<RexNode, RexNode> evaluatedResults = new HashMap<>();
+    AtomicReference<Set<RexNode>> callsToEval = new AtomicReference<>(new HashSet<>());
+
+    // Recursively evaluate all the calls with Literal only operands from bottom up
+    do {
+      callsToEval.get().clear();
+      // Traverse the relational expression using a RexShuttle visitor
+      rootRelNode.accept(new RexShuttle() {
+        @Override
+        public RexNode visitCall(RexCall call) {
+          call.operands.forEach(operand -> {
+            if (operand instanceof RexCall) {
+              visitCall((RexCall) operand);
+            }
+          });
+          // Check if all operands are RexLiteral
+          if (call.operands.stream().allMatch(operand -> operand instanceof RexLiteral)) {
+            // If all operands are literals or already evaluated, this call can be evaluated
+            if (!evaluatedResults.containsKey(call)) {
+              callsToEval.get().add(call);
+            }
+          }
+          // Return the call to continue traversal
+          return call;
```

Comment:
Can we directly return the evaluated result here? I think it can directly replace the node

Reply:
done.